catkin_ws/src/kpr_testing/src/valves.py

```diff
@@ -0,0 +1,163 @@
+#!/usr/bin/python
+
+import rospy, termios, sys, contextlib, os
+import numpy as np
+from ur_msgs.srv import *
+from ur_msgs.msg import *
+
+header = """
+Check the status of I/O-pins
+\t-^D to exit
+\t-i to change income settings
+\t-p to send io states
+\t-t to toggle service succes
+"""
+
+def display():
+	os.system('clear')
+	print header
+#Making table
+	tableHeader = "Pin     |  State     Fun     "
+	print tableHeader
+	print "-"*len(tableHeader)
+	print "VacSens | ", info[VacSens][0], "  ", info[VacSens][1]
+	print "Gripper | ", info[Gripper][0], "  ", info[Gripper][1]
+
+#These IO pins are used
+VacSens = "DI1"
+Gripper = "DO2"
+
+#Set succes response
+response = True
+
+#dictionary of pins
+info = {VacSens:["-","     -"], Gripper:["-","     -"]}
+
+#needed for the "pressing keys"
+@contextlib.contextmanager
+def raw_mode(file):
+    old_attrs = termios.tcgetattr(file.fileno())
+    new_attrs = old_attrs[:]
+    new_attrs[3] = new_attrs[3] & ~(termios.ECHO | termios.ICANON)
+    try:
+        termios.tcsetattr(file.fileno(), termios.TCSADRAIN, new_attrs)
+        yield
+    finally:
+        termios.tcsetattr(file.fileno(), termios.TCSADRAIN, old_attrs)
+
+def request(req):
+	global response
+	display()
+#Checking if changes wil be made to used pins 
+	if req.fun == 1:
+		pin = 'DO' + str(req.pin)
+	elif req.fun == 4:
+		pin = 'AO' + str(req.pin)
+	else:
+		rospy.logerr("Writing to unused pin")
+		return SetIOResponse(False)
+	if pin not in info:
+		rospy.logerr("Writing to unused pin")
+		return SetIOResponse(False)
+#Change Dictionary "info" with correct data
+	info[pin][0] = req.state
+	if req.fun == 1:
+		info[pin][1] = "Digital Out"
+	else:
+		info[pin][1] = "Analog Out"
+
+	display()
+	res = response	
+	return SetIOResponse(res)
+		
+def get_message():
+	msg = IOStates()
+	for pin in info:
+		if info[pin][1] == "Digital Out":
+			msg.digital_out_states.append(Digital(int(pin[2]), bool(info[pin][0])))
+		elif info[pin][1] == "Analog Out":
+			msg.analog_out_states.append(Analog(int(pin[2]), float(info[pin][0])))
+		elif info[pin][1] == "Digital In":
+			msg.digital_in_states.append(Digital(int(pin[2]), bool(info[pin][0])))
+		elif info[pin][1] == "Analog In":
+			msg.analog_in_states.append(Analog(int(pin[2]), float(info[pin][0])))
+	return msg
+
+def processKey(key, pub):
+	global response
+	if key == chr(116): #t
+		response = not response
+		display()
+		print "Your response is now: ", response
+		return
+	if key == chr(112): #p
+		pub.publish(get_message())
+		display()
+		print "you have published the data"
+		return
+	if key == chr(105): #i
+		display()
+		print "Changing the input settings"
+		income()
+		return
+	rospy.logwarn('unknown function %s', key)
+
+def income():
+	try:
+		fun = int(raw_input("What is the fun? "))
```

Comment:
You might want to give the user clues about the two possible values he could enter here